File: src/services/auth.service.ts
import bcrypt from 'bcryptjs';
import jwt, { SignOptions } from 'jsonwebtoken';
import { v4 as uuidv4 } from 'uuid';
import crypto from 'crypto';
import pool, { query, queryOne, execute } from '../config/db';
import env from '../config/env';
import { formatUser } from '../utils/formatAd';

// ─── Token helpers ────────────────────────────────────────────────────────────

function signAccessToken(userId: string, email: string): string {
  return jwt.sign(
    { sub: userId, email },
    env.JWT_ACCESS_SECRET,
    { expiresIn: env.JWT_ACCESS_EXPIRES_IN as SignOptions['expiresIn'] }
  );
}

function signRefreshToken(userId: string): string {
  return jwt.sign(
    { sub: userId },
    env.JWT_REFRESH_SECRET,
    { expiresIn: env.JWT_REFRESH_EXPIRES_IN as SignOptions['expiresIn'] }
  );
}

function hashToken(token: string): string {
  return crypto.createHash('sha256').update(token).digest('hex');
}

/** Persist a new refresh token and return the raw JWT string. */
async function issueRefreshToken(userId: string): Promise<string> {
  const rawToken = signRefreshToken(userId);
  const hash     = hashToken(rawToken);
  const id       = uuidv4();

  // Expiry: parse env string (e.g. "30d") to ms
  const days  = parseInt(env.JWT_REFRESH_EXPIRES_IN.replace(/\D/g, ''), 10) || 30;
  const expAt = new Date(Date.now() + days * 24 * 60 * 60 * 1000);

  await execute(
    `INSERT INTO refresh_tokens (id, user_id, token_hash, expires_at, created_at) VALUES (?, ?, ?, ?, ?)`,
    [id, userId, hash, expAt, new Date()]
  );

  return rawToken;
}

// ─── Service ─────────────────────────────────────────────────────────────────

import { emailService } from './email.service';

// ... (existing signAccessToken, signRefreshToken, hashToken, issueRefreshToken remain unchanged)

export const authService = {
  // ── Register ────────────────────────────────────────────────────────────────
  async register(name: string, email: string, phone: string, password: string) {
    const emailExists = await queryOne('SELECT id FROM users WHERE email = ?', [email.toLowerCase()]);
    if (emailExists) throw Object.assign(new Error('An account with this email already exists.'), { status: 409, code: 'email_taken' });

    const phoneExists = await queryOne('SELECT id FROM users WHERE phone = ?', [phone]);
    if (phoneExists) throw Object.assign(new Error('This phone number is already registered.'), { status: 409, code: 'phone_taken' });

    const hash = await bcrypt.hash(password, 12);
    const id   = uuidv4();
    const verificationToken = crypto.randomBytes(32).toString('hex');

    const now = new Date();
    await execute(
      `INSERT INTO users (id, name, email, phone, password_hash, verification_token, created_at, updated_at) VALUES (?, ?, ?, ?, ?, ?, ?, ?)`,
      [id, name.trim(), email.toLowerCase(), phone, hash, verificationToken, now, now]
    );

    // Send verification link
    emailService.sendVerificationLink(email.toLowerCase(), verificationToken).catch(console.error);

    const user = await queryOne('SELECT * FROM users WHERE id = ?', [id]);

    const accessToken  = signAccessToken(id, email.toLowerCase());
    const refreshToken = await issueRefreshToken(id);

    return { user: formatUser(user), access_token: accessToken, refresh_token: refreshToken };
  },

  // ── Verify Email ─────────────────────────────────────────────────────────────
  async verifyEmail(token: string) {
    const user = await queryOne<any>('SELECT * FROM users WHERE verification_token = ?', [token]);
    if (!user) throw Object.assign(new Error('Invalid or expired verification token.'), { status: 400, code: 'invalid_token' });

    if (user.is_verified) return { message: 'Email already verified.' };

    await execute('UPDATE users SET is_verified = 1, verification_token = NULL, updated_at = ? WHERE id = ?', [new Date(), user.id]);
    return { message: 'Email verified successfully.' };
  },

  // ── Login ────────────────────────────────────────────────────────────────────
  async login(email: string, password: string) {
    const user = await queryOne<any>('SELECT * FROM users WHERE email = ?', [email.toLowerCase()]);
    if (!user) throw Object.assign(new Error('Incorrect email or password.'), { status: 401, code: 'invalid_credentials' });

    const valid = await bcrypt.compare(password, user.password_hash);
    if (!valid) throw Object.assign(new Error('Incorrect email or password.'), { status: 401, code: 'invalid_credentials' });

    // Check if verified
    if (!user.is_verified) {
      throw Object.assign(new Error('Please verify your email address before logging in.'), { status: 403, code: 'not_verified' });
    }

    const accessToken  = signAccessToken(user.id, user.email);
    const refreshToken = await issueRefreshToken(user.id);

    return { user: formatUser(user), access_token: accessToken, refresh_token: refreshToken };
  },

  // ── Forgot Password ──────────────────────────────────────────────────────────
  async forgotPassword(email: string) {
    const user = await queryOne<any>('SELECT * FROM users WHERE email = ?', [email.toLowerCase()]);
    if (!user) throw Object.assign(new Error('No account found with this email.'), { status: 404, code: 'not_found' });

    const token = crypto.randomBytes(32).toString('hex');
    const expires = new Date(Date.now() + 15 * 60 * 1000); // 15 minutes

    await execute(
      'UPDATE users SET reset_password_token = ?, reset_password_expires = ?, updated_at = ? WHERE id = ?',
      [token, expires, new Date(), user.id]
    );

    await emailService.sendPasswordResetLink(email.toLowerCase(), token);
    return { message: 'Password reset link sent to your email.' };
  },

  // ── Reset Password ───────────────────────────────────────────────────────────
  async resetPassword(token: string, newPassword: string) {
    const user = await queryOne<any>('SELECT * FROM users WHERE reset_password_token = ?', [token]);
    if (!user) throw Object.assign(new Error('Invalid or expired reset token.'), { status: 400, code: 'invalid_token' });

    if (new Date(user.reset_password_expires) < new Date()) {
      throw Object.assign(new Error('Reset link has expired.'), { status: 400, code: 'expired_token' });
    }

    const hash = await bcrypt.hash(newPassword, 12);
    await execute(
      'UPDATE users SET password_hash = ?, reset_password_token = NULL, reset_password_expires = NULL, updated_at = ? WHERE id = ?',
      [hash, new Date(), user.id]
    );

    return { message: 'Password has been reset successfully.' };
  },

  // ── Refresh ──────────────────────────────────────────────────────────────────
  async refresh(rawRefreshToken: string) {
    // ... (logic remains same)
    // Verify JWT signature first
    let payload: any;
    try {
      payload = jwt.verify(rawRefreshToken, env.JWT_REFRESH_SECRET);
    } catch {
      throw Object.assign(new Error('Invalid or expired refresh token.'), { status: 401, code: 'token_invalid' });
    }

    const hash = hashToken(rawRefreshToken);
    const stored = await queryOne<any>(
      'SELECT * FROM refresh_tokens WHERE token_hash = ? AND revoked = 0',
      [hash]
    );

    if (!stored || new Date(stored.expires_at) < new Date()) {
      throw Object.assign(new Error('Refresh token is invalid or has been revoked.'), { status: 401, code: 'token_invalid' });
    }

    // Revoke old token (rotation)
    await execute('UPDATE refresh_tokens SET revoked = 1 WHERE token_hash = ?', [hash]);

    const user = await queryOne<any>('SELECT * FROM users WHERE id = ?', [payload.sub]);
    if (!user) throw Object.assign(new Error('User not found.'), { status: 401, code: 'not_authenticated' });

    const newAccessToken  = signAccessToken(user.id, user.email);
    const newRefreshToken = await issueRefreshToken(user.id);

    return { access_token: newAccessToken, refresh_token: newRefreshToken };
  },

  // ── Logout ───────────────────────────────────────────────────────────────────
  async logout(rawRefreshToken: string) {
    const hash = hashToken(rawRefreshToken);
    await execute('UPDATE refresh_tokens SET revoked = 1 WHERE token_hash = ?', [hash]);
  },

  // ── Get current user ─────────────────────────────────────────────────────────
  async getMe(userId: string) {
    const user = await queryOne<any>('SELECT * FROM users WHERE id = ?', [userId]);
    if (!user) throw Object.assign(new Error('User not found.'), { status: 404, code: 'not_found' });
    return formatUser(user);
  },
};

File: EasyPhones.Application/Services/AuthService.cs
using System.Net;
using System.Security.Cryptography;
using System.Text;
using EasyPhones.Application.Dtos.AuthDtos;
using EasyPhones.Application.Exceptions;
using EasyPhones.Domain.Entities;
using EasyPhones.Domain.Repositories;
using Microsoft.Extensions.Configuration;
using Microsoft.IdentityModel.Tokens;
using System.IdentityModel.Tokens.Jwt;
using System.Security.Claims;
using BCrypt.Net;

namespace EasyPhones.Application.Services;

public class AuthService
{
    private readonly IUserRepository _users;
    private readonly IRefreshTokenRepository _tokens;
    private readonly IEmailService _email;
    private readonly IConfiguration _config;

    public AuthService(
        IUserRepository users,
        IRefreshTokenRepository tokens,
        IEmailService email,
        IConfiguration config)
    {
        _users = users;
        _tokens = tokens;
        _email = email;
        _config = config;
    }

    // ── Register ─────────────────────────────────────────────────────────────
    public async Task<AuthResultDto> RegisterAsync(RegisterDto dto, CancellationToken ct)
    {
        if (dto.Password != dto.ConfirmPassword)
            throw ApiException.Validation("Passwords do not match.");

        var existing = await _users.GetByEmailAsync(dto.Email.ToLower(), ct);
        if (existing is not null)
            throw ApiException.Conflict("An account with this email already exists.", "email_taken");

        var phoneExisting = await _users.GetByPhoneAsync(dto.Phone, ct);
        if (phoneExisting is not null)
            throw ApiException.Conflict("This phone number is already registered.", "phone_taken");

        var hash = BCrypt.Net.BCrypt.HashPassword(dto.Password, 12);
        var verificationToken = GenerateSecureToken();

        var user = new User
        {
            Id = Guid.NewGuid().ToString(),
            Name = dto.Name.Trim(),
            Email = dto.Email.ToLower(),
            Phone = dto.Phone,
            PasswordHash = hash,
            VerificationToken = verificationToken,
            CreatedAt = DateTime.UtcNow,
            UpdatedAt = DateTime.UtcNow
        };

        await _users.AddAsync(user, ct);

        _ = _email.SendVerificationLinkAsync(user.Email, verificationToken);

        var (accessToken, refreshToken) = await IssueTokenPairAsync(user, ct);

        return new AuthResultDto
        {
            Message = "Email sent, please verify your email to proceed.",
            User = MapUser(user),
            AccessToken = accessToken,
            RefreshToken = refreshToken
        };
    }

    // ── Verify Email ──────────────────────────────────────────────────────────
    public async Task<string> VerifyEmailAsync(string token, CancellationToken ct)
    {
        var user = await _users.GetByVerificationTokenAsync(token, ct)
            ?? throw ApiException.Validation("Invalid or expired verification token.", "invalid_token");

        if (user.IsVerified) return "Email already verified.";

        user.IsVerified = true;
        user.VerificationToken = null;
        user.UpdatedAt = DateTime.UtcNow;
        await _users.UpdateAsync(user, ct);

        return "Email verified successfully.";
    }

    // ── Login ─────────────────────────────────────────────────────────────────
    public async Task<AuthResultDto> LoginAsync(LoginDto dto, CancellationToken ct)
    {
        var user = await _users.GetByEmailAsync(dto.Email.ToLower(), ct)
            ?? throw ApiException.Unauthorized("Incorrect email or password.", "invalid_credentials");

        if (!BCrypt.Net.BCrypt.Verify(dto.Password, user.PasswordHash))
            throw ApiException.Unauthorized("Incorrect email or password.", "invalid_credentials");

        if (!user.IsVerified)
            throw new ApiException("Please verify your email address before logging in.", HttpStatusCode.Forbidden, "not_verified");

        var (accessToken, refreshToken) = await IssueTokenPairAsync(user, ct);

        return new AuthResultDto
        {
            Message = "Login successful",
            User = MapUser(user),
            AccessToken = accessToken,
            RefreshToken = refreshToken
        };
    }

    // ── Forgot Password ───────────────────────────────────────────────────────
    public async Task<string> ForgotPasswordAsync(ForgotPasswordDto dto, CancellationToken ct)
    {
        var user = await _users.GetByEmailAsync(dto.Email.ToLower(), ct)
            ?? throw ApiException.NotFound("No account found with this email.");

        var token = GenerateSecureToken();
        user.ResetPasswordToken = token;
        user.ResetPasswordExpires = DateTime.UtcNow.AddMinutes(15);
        user.UpdatedAt = DateTime.UtcNow;
        await _users.UpdateAsync(user, ct);

        await _email.SendPasswordResetLinkAsync(dto.Email.ToLower(), token);
        return "Password reset link sent to your email.";
    }

    // ── Reset Password ────────────────────────────────────────────────────────
    public async Task<string> ResetPasswordAsync(ResetPasswordDto dto, CancellationToken ct)
    {
        var user = await _users.GetByResetTokenAsync(dto.Token, ct)
            ?? throw ApiException.Validation("Invalid or expired reset token.", "invalid_token");

        if (user.ResetPasswordExpires < DateTime.UtcNow)
            throw ApiException.Validation("Reset link has expired.", "expired_token");

        user.PasswordHash = BCrypt.Net.BCrypt.HashPassword(dto.Password, 12);
        user.ResetPasswordToken = null;
        user.ResetPasswordExpires = null;
        user.UpdatedAt = DateTime.UtcNow;
        await _users.UpdateAsync(user, ct);

        return "Password has been reset successfully.";
    }

    // ── Refresh ───────────────────────────────────────────────────────────────
    public async Task<TokenResultDto> RefreshAsync(string rawRefreshToken, CancellationToken ct)
    {
        string userId;
        try
        {
            var principal = ValidateRefreshToken(rawRefreshToken);
            userId = principal.FindFirst(JwtRegisteredClaimNames.Sub)?.Value
                ?? throw new SecurityTokenException();
        }
        catch
        {
            throw ApiException.Unauthorized("Invalid or expired refresh token.", "token_invalid");
        }

        var hash = HashToken(rawRefreshToken);
        var stored = await _tokens.GetByHashAsync(hash, ct);

        if (stored is null || stored.Revoked || stored.ExpiresAt < DateTime.UtcNow)
            throw ApiException.Unauthorized("Refresh token is invalid or has been revoked.", "token_invalid");

        await _tokens.RevokeByHashAsync(hash, ct);

        var user = await _users.GetByIdAsync(userId, ct)
            ?? throw ApiException.Unauthorized("User not found.", "not_authenticated");

        var (newAccessToken, newRefreshToken) = await IssueTokenPairAsync(user, ct);
        return new TokenResultDto { AccessToken = newAccessToken, RefreshToken = newRefreshToken };
    }

    // ── Logout ────────────────────────────────────────────────────────────────
    public async Task LogoutAsync(string? rawRefreshToken, CancellationToken ct)
    {
        if (string.IsNullOrEmpty(rawRefreshToken)) return;
        var hash = HashToken(rawRefreshToken);
        await _tokens.RevokeByHashAsync(hash, ct);
    }

    // ── Get Me ────────────────────────────────────────────────────────────────
    public async Task<UserDto> GetMeAsync(string userId, CancellationToken ct)
    {
        var user = await _users.GetByIdAsync(userId, ct)
            ?? throw ApiException.NotFound("User not found.");
        return MapUser(user);
    }

    // ── Helpers ───────────────────────────────────────────────────────────────
    private async Task<(string AccessToken, string RefreshToken)> IssueTokenPairAsync(User user, CancellationToken ct)
    {
        var accessToken = SignAccessToken(user.Id, user.Email);
        var rawRefresh = SignRefreshToken(user.Id);
        var hash = HashToken(rawRefresh);

        var days = ParseDays(_config["JwtOptions:RefreshExpiresIn"] ?? "30d");
        var refreshEntity = new RefreshToken
        {
            Id = Guid.NewGuid().ToString(),
            UserId = user.Id,
            TokenHash = hash,
            ExpiresAt = DateTime.UtcNow.AddDays(days),
            CreatedAt = DateTime.UtcNow,
            UpdatedAt = DateTime.UtcNow
        };
        await _tokens.AddAsync(refreshEntity, ct);

        return (accessToken, rawRefresh);
    }

    private string SignAccessToken(string userId, string email)
    {
        var key = new SymmetricSecurityKey(Encoding.UTF8.GetBytes(_config["JwtOptions:SecretKey"]!));
        var creds = new SigningCredentials(key, SecurityAlgorithms.HmacSha256);
        var expires = DateTime.UtcNow.Add(ParseExpiry(_config["JwtOptions:AccessExpiresIn"] ?? "15m"));

        var token = new JwtSecurityToken(
            issuer: _config["JwtOptions:Issuer"],
            audience: _config["JwtOptions:Audience"],
            claims: [
                new Claim(JwtRegisteredClaimNames.Sub, userId),
                new Claim(JwtRegisteredClaimNames.Email, email),
                new Claim(JwtRegisteredClaimNames.Jti, Guid.NewGuid().ToString())
            ],
            expires: expires,
            signingCredentials: creds
        );
        return new JwtSecurityTokenHandler().WriteToken(token);
    }

    private string SignRefreshToken(string userId)
    {
        var key = new SymmetricSecurityKey(Encoding.UTF8.GetBytes(_config["JwtOptions:RefreshSecretKey"]!));
        var creds = new SigningCredentials(key, SecurityAlgorithms.HmacSha256);
        var days = ParseDays(_config["JwtOptions:RefreshExpiresIn"] ?? "30d");

        var token = new JwtSecurityToken(
            claims: [new Claim(JwtRegisteredClaimNames.Sub, userId)],
            expires: DateTime.UtcNow.AddDays(days),
            signingCredentials: creds
        );
        return new JwtSecurityTokenHandler().WriteToken(token);
    }

    private ClaimsPrincipal ValidateRefreshToken(string token)
    {
        var key = new SymmetricSecurityKey(Encoding.UTF8.GetBytes(_config["JwtOptions:RefreshSecretKey"]!));
        return new JwtSecurityTokenHandler().ValidateToken(token, new TokenValidationParameters
        {
            ValidateIssuerSigningKey = true,
            IssuerSigningKey = key,
            ValidateIssuer = false,
            ValidateAudience = false,
            ClockSkew = TimeSpan.Zero
        }, out _);
    }

    private static string HashToken(string token)
    {
        var bytes = SHA256.HashData(Encoding.UTF8.GetBytes(token));
        return Convert.ToHexString(bytes).ToLower();
    }

    private static string GenerateSecureToken()
        => Convert.ToHexString(RandomNumberGenerator.GetBytes(32)).ToLower();

    private static int ParseDays(string s)
    {
        if (s.EndsWith('d') && int.TryParse(s[..^1], out var d)) return d;
        return 30;
    }

    private static TimeSpan ParseExpiry(string s)
    {
        if (s.EndsWith('m') && int.TryParse(s[..^1], out var m)) return TimeSpan.FromMinutes(m);
        if (s.EndsWith('h') && int.TryParse(s[..^1], out var h)) return TimeSpan.FromHours(h);
        if (s.EndsWith('d') && int.TryParse(s[..^1], out var d)) return TimeSpan.FromDays(d);
        return TimeSpan.FromMinutes(15);
    }

    public static UserDto MapUser(User u, string? avatarUrl = null) => new()
    {
        Id = u.Id,
        Name = u.Name,
        Email = u.Email,
        Phone = u.Phone,
        AvatarUrl = avatarUrl ?? u.AvatarUrl,
        CreatedAt = u.CreatedAt.ToString("O")
    };
}
